pkg/local/sync.go

```diff
@@ -73,28 +73,22 @@ func (s *SyncManager) Sync(rootPath string, remote *uri.URI, changeSet <-chan *C
 	s.progressBar.Start()
 	defer s.progressBar.Stop()
 
-	ch := make(chan bool, s.maxParallelism)
-	for i := 0; i < s.maxParallelism; i++ {
-		ch <- true
-	}
-
 	wg, ctx := errgroup.WithContext(s.ctx)
-	done := false
-	for change := range changeSet {
-		<-ch // block until we have a slot
-		c := change
-		select {
-		case <-ctx.Done():
-			done = true
-		default:
-			wg.Go(func() error {
-				return s.apply(ctx, rootPath, remote, c)
-			})
-		}
-		ch <- true // release
-		if done {
-			break
-		}
+	for i := 0; i < s.maxParallelism; i++ {
```

Comment:
I don't understand why use an errgroup and then also try to control parallelism with a chan. Why not wg.SetLimit to parallelism and then just wg.Go each change in a simple loop?

Reply:
Because SetLimit is blocking, and using TryGo will just make us iterate on a loop unnecessarily

Reply:
I understand that SetLimit will make Go block when there is no goroutine ready to receive the change.  I don't understand why this is a bad thing: I have a changeSet, it has a capacity.  All this goroutine does is load parallelism goroutines and each of them tries to read changeSet.  If we called wg.Go() directly, it would occasionally block - until the next parallelism slot were opened.  I don't think you'll be able to measure a difference in performance, just I'd like to have a loop that looks like this:

```go
wg.SetLimit(parallelism)
for change:= range changeSet {
	c := change
	wg.Go(func() error { return s.apply(ctx, rootPath, remote, c) }
}
if err := wg.Wait(); err != nil {
	return err
}
// ...
```

Reply:
Thanks - I understand now what you mean - modified accordingly